### DISCUSSION TITLE
VS Code extension

### DISCUSSION THREAD

**Discussion by felipecrs** [twpayne/chezmoi#1737] [Ideas]
## Is your feature request related to a problem? Please describe.

Mix and matching chezmoi templates with the various language types it accepts is cumbersome in VS Code. An extension would help in that.

References:

- https://github.com/chezmoi/dotfiles/pull/8
- https://github.com/twpayne/chezmoi/discussions/1426

## Describe the solution you'd like

A chezmoi extension for VS Code, with the chezmoi binary providing a language server for the extension to pull information from (such as list of variables, allowed functions, and so). Similar to how deno bundles the language server and provides it with the `deno lsp` command (all the data is exchanged through stdin and stdout).

Some references on writing the language server in Go:

- https://github.com/sourcegraph/go-lsp

Other references:

- https://lsif.dev/
- https://langserver.org/

The good thing is that, with a LSP, it's not difficult for other editors to provide the same capabilities.

I suspect that a lot of code can be reused from:

- https://github.com/jinliming2/vscode-go-template
- https://github.com/tim-koehler/Helm-Intellisense

## Features

This is a list of features I can think the extension can provide (I will add more as I find):

- Templates syntax highlight on different files extension, such as Shell Script
- Apply on save?
- Intellisense on templates
  - Hove on variables to know their actual values - Helm Intellisense does it for values defined in `values.yaml`
  - Autocomplete variables, functions
  - Docs for functions
  - Provide intellisense for templates without removing the original language intellisense (such as mixing Shell Script with Go Template)
- Real time template analyze (linting like), such as type-check which would happen during `execute-template`
- Use SCM API to provide diff, changes (git-like) for files in home

**Comment by BasixKOR** [comment]
Here are some requirements I can think of right now:
 * Resolve [source code attributes](https://www.chezmoi.io/docs/reference/#source-state-attributes) to help VS Code identify the language of the target file 
 * [Syntax highlighting for templates](https://github.com/jinliming2/vscode-go-template)
 
There is a syntax highlighter for `text/template` for VS Code, perhaps we could leverage that.
 
Some guides that could be helpful for implementors:
 * https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#injection-grammars
 * https://code.visualstudio.com/api/language-extensions/embedded-languages (not same but close)

**Comment by twpayne** [comment]
As this is rather open-ended, I'll convert this to a discussion.
